core/src/error/mod.rs

```diff
@@ -1303,6 +1311,26 @@ impl IntoDiagnostics<FileId> for EvalError {
                         "But it has been applied to an argument which doesn't match any of those patterns".to_owned(),
                     ])]
             }
+            EvalError::NonExhaustiveMatch { value, pos } => {
+                let mut labels = Vec::new();
+
+                if let Some(span) = pos.into_opt() {
+                    labels.push(primary(&span).with_message("in this match expression"));
+                }
+
+                labels.push(
+                    secondary_term(&value, files)
+                        .with_message("this value doesn't match any branch"),
+                );
+
+                vec![Diagnostic::error()
+                    .with_message("non-exhaustive pattern matching")
+                    .with_labels(labels)
+                    .with_notes(vec![
+                        format!("This match expression isn't exhaustive"),
+                        "It has been applied to an argument which doesn't match any of the branches".to_owned(),
+                    ])]
```

Comment:
```suggestion
                vec![Diagnostic::error()
                    .with_message("unmatched pattern")
                    .with_labels(labels)
                    ])]
```
I think the notes are redundant given the labels, and I also think "unmatched" is better than "non-exhaustive" here, because a match expression is non-exhaustive if there *exists* a value which isn't matched whereas this runtime error occurs if we've *encountered* a value which isn't matched. (Which of course implies the existence of one, but I think we can use the more specific error message)